paper/content/well-typed-holes.tex

```diff
@@ -1,28 +1,59 @@
 \subsection{Well-Typed Holes}
 \label{sec:design:holes}
 
-Typed-holes are a useful feature in type-driven programming~\cite{DBLP:journals/pacmpl/OmarVCH19}, and supports the \emph{dropping in} of a special term, the hole, into a program such that during elaboration the type of that term will be synthesised.
-User's can inspect the type of a hole and will be presented with the local context that will, hopefully, show useful terms to help fill the hole.
+Holes are a special kind of placeholder that programmers can use for parts of
+the program they have not yet written.
+%
+In a typed language, each hole will be assigned a type based on the context it
+is used in.\gallais{context as in $C[\cdot]$, not $\Gamma$. Better phrasing?}
+
+Special language support (the ability to inspect, refine, compute with,
+or fill an existing hole with an adequately typed term)
+enables \emph{type-driven programming}~\cite{DBLP:journals/pacmpl/OmarVCH19},
+a practice by which the user interactively builds the eventual finished
+program via a dialogue with the compiler.
+
+Barebones language support should at least include the ability to inspect the
+type of a hole and the local context it appears in, as well as the ability to
+safely evaluate programs that still contain holes.
+\todo{Implement \texttt{instantiate}? Shouldn't take too long.}
+
+
+
+\Idris{}'s elaborator lifts holes to top-level declarations with no associated
+definition as it encounters them.
+%
+As a consequence of this design choice, users cannot mention the same hole in
+different places to state their intention that these yet unwritten terms ought
+to be the same.
```

Comment:
Actually, I have remembered that this is not exactly true. You _can_ reuse a hole in Idris, but not as a hole, but by the hole's name. As an example:

```idris
plus : Nat -> Nat -> Nat
plus = ?rhs

double : Nat -> Nat
double x = rhs x x
```

Successfully type-checks with the following goal left:

```
Holes

- + "Main.rhs" [P]
 `-- Nat -> Nat -> Nat
```

Reply:
I have address this, hopefully, satisfactorily.

Reply:
I think this is an accident of the implementation rather than a voluntary feature.

The hole is still very much placed in one position (and it is from that position that we infer its context)
instead of being present in multiple places (and we would then infer a context that is valid for both positions).

Reply:
> I think this is an accident of the implementation rather than a voluntary feature.

I agree! but it is something we should mention, and we should just copy and paste this in

> The hole is still very much placed in one position (and it 
> is from that position that we infer its context)
> instead of being present in multiple places (and we would 
> then infer a context that is valid for both positions).